libwasmvm/src/error/go.rs

```diff
@@ -15,77 +15,185 @@ use crate::memory::UnmanagedVector;
-    /// If it is GoResult::User the error message will be returned to the contract.
+impl GoError {
+    /// This converts a GoError to a `Result<(), BackendError>`, using a fallback error message for some cases.
+    /// If it is GoError::User the error message will be returned to the contract.
     /// Otherwise, the returned error will trigger a trap in the VM and abort contract execution immediately.
     ///
-    /// Safety: this reads data from an externally provided buffer and assumes valid utf-8 encoding
-    /// Only call if you trust the code that provides `error_msg` to be correct.
-    pub unsafe fn into_ffi_result<F>(
+    /// This reads data from an externally provided `UnmanagedVector` and assumes UFT-8 encoding. To protect
+    /// against invalid UTF-8 data, a lossy conversion to string is used. The data is limited to 8KB in order
+    /// to protect against long externally generated error messages.
+    ///
+    /// The `error_msg` is always consumed here and must not be used afterwards.
+    pub unsafe fn into_result<F>(
         self,
         error_msg: UnmanagedVector,
-        default: F,
+        default_error_msg: F,
     ) -> Result<(), BackendError>
     where
-        F: Fn() -> String,
+        F: FnOnce() -> String,
     {
-        let read_error_msg = || -> String {
-            match error_msg.consume() {
-                Some(data) => String::from_utf8_lossy(&data).into(),
-                None => default(),
+        const MAX_ERROR_LEN: usize = 8 * 1024;
+
+        // We destruct the UnmanagedVector here, no matter if we need the data.
+        let error_msg = error_msg.consume();
+
+        let build_error_msg = || -> String {
+            match error_msg {
+                Some(mut data) => {
+                    data.truncate(MAX_ERROR_LEN);
+                    String::from_utf8_lossy(&data).into()
+                }
+                None => default_error_msg(),
             }
         };
 
         match self {
-            GoResult::Ok => Ok(()),
-            GoResult::Panic => Err(BackendError::foreign_panic()),
-            GoResult::BadArgument => Err(BackendError::bad_argument()),
-            GoResult::OutOfGas => Err(BackendError::out_of_gas()),
-            GoResult::Other => Err(BackendError::unknown(read_error_msg())),
-            GoResult::User => Err(BackendError::user_err(read_error_msg())),
+            // Success
+            GoError::None => Ok(()),
+            // Errors with direct counterpart
+            GoError::Panic => Err(BackendError::foreign_panic()),
+            GoError::BadArgument => Err(BackendError::bad_argument()),
+            GoError::OutOfGas => Err(BackendError::out_of_gas()),
+            GoError::User => Err(BackendError::user_err(build_error_msg())),
+            // Everything else goes into unknown
+            GoError::CannotSerialize | GoError::Other => {
```

Comment:
What is the purpose of this new GoError variant if we fold it into 
BackendError::unknown?

Reply:
We have more information on the Rust side. From there we can improve BackendError to make it more granular as a next step. But since all cases except `BackendError::UserErr` are handled the same way, it's not so important.